### PR TITLE
[Feature] Support DPO, ORPO and Reward Model

### PR DESCRIPTION
**Pull request by RangiLyu**
## DPO, ORPO, and Reward Model Training

image

### Feature List

- Efficient training by utilizing packed chosen/rejected pairs when `flash_attention` is used. Significantly minimizes GPU memory wastage due to padding tokens in batch training.
- Unified dataset format across DPO, ORPO, and Reward Model.
- Support for QLora training
- Support for sequence parallel training

### Tasks Pending

- [x] Documentation
- [x] Convert reward model to huggingface format 
- [x] Validationdataset

### CODE REVIEW

File: xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    llm=dict(
+        type=AutoModelForCausalLM.from_pretrained,
+        pretrained_model_name_or_path=pretrained_model_name_or_path,
+        trust_remote_code=True))
+
+#######################################################################
+#                      PART 3  Dataset & Dataloader                   #
+#######################################################################
+train_dataset = dict(
+    type=build_preference_dataset,
```
Comment (hhaAndroid) [inline xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py:82]: 只是换成了 jsonl 格式的数据应该不需要再加一个配置吧？ 感觉可以以注释形式加到其他配置里面就行，这个配置就可以不要了

File: xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+from xtuner.engine.runner import TrainLoop
+from xtuner.model.reward import RewardModel
+
+#######################################################################
+#                          PART 1  Settings                           #
+#######################################################################
+# Model
+pretrained_model_name_or_path = 'internlm/internlm2-chat-1_8b-sft'
+use_varlen_attn = True
+reward_token_id = 92527
```
Comment (hhaAndroid) [inline xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py]: 需要加个注释

File: xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_ultrafeedback.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    tokenizer=tokenizer,
+    max_length=max_length,
+    dataset_map_fn=orpo_dpo_mix_40k_map_fn,
+    is_dpo=False,
+    is_reward=True,
+    reward_token_id=reward_token_id,
+    num_proc=32,
+    use_varlen_attn=use_varlen_attn,
+    shuffle_before_pack=True,
+    seed=42,
```
Comment (hhaAndroid) [inline xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_ultrafeedback.py]: 这里如果强行设置 seed，那如果 train.py 里面也设置了，是不是无法生效？

File: xtuner/configs/orpo/internlm/internlm2_chat_1_8b_orpo_full.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    padding_side='right')
+
+model = dict(
+    type=ORPO,
+    use_varlen_attn=use_varlen_attn,
+    llm=dict(
+        type=AutoModelForCausalLM.from_pretrained,
+        pretrained_model_name_or_path=pretrained_model_name_or_path,
+        trust_remote_code=True),
+    lora=dict(
```
Comment (hhaAndroid) [inline xtuner/configs/orpo/internlm/internlm2_chat_1_8b_orpo_full.py]: 这个要删掉，因为这不是 full

File: xtuner/model/modules/dispatch/qwen2.py [hunk 5]
```diff
@@ -4,8 +4,7 @@
 from typing import Optional
 
 import torch
-import torch.distributed as dist
-from mmengine import MessageHub
+from mmengine import MessageHub, dist
```
Comment (HIT-cwh) [inline xtuner/model/modules/dispatch/qwen2.py]: mmengine的 dist 接口跟 torch.distributed 的接口是不一致的，例如 mmengine.dist 没有dist.is_initialized()这个方法。

File: xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py (hunk 1 above)
Comment (RangiLyu) [reply xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_varlenattn_jsonl_dataset.py:82]: 有个实际的例子对用户来说会简单很多，文档也可以直接链接过来

File: xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_ultrafeedback.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+use_varlen_attn = False
+reward_token_id = 92527  # use [UNUSED_TOKEN_130] as reward token
+loss_type = 'focal'
+penalty_type = 'log_barrier'
+
+# Data
+max_length = 2048
+
+# Scheduler & Optimizer
+batch_size = 4  # per_device
```
Comment (hhaAndroid) [inline xtuner/configs/reward_model/internlm/internlm2_chat_1_8b_reward_full_ultrafeedback.py:32]: 其他我看都是 1，为啥这个是 4？

**Review by hhaAndroid — APPROVED** [review]
